### PR TITLE
Add explicit weight decay

### PR DESCRIPTION
**Pull request by bencherian**
Adds explicit support for weight decay as per request in #2784.
This enables better optimization behavior for adaptive gradient methods (e.g., adagrad, adam) and easier hyperparameter tuning all techniques outside of vanilla SGD due to improved separation between learning rate and regularization terms. In order to enable weight decay schemes discussed in the [paper](https://arxiv.org/abs/1711.05101) which highlights this issue, the user is able to specify the weight decay using a schedule.

### CODE REVIEW

**Comment by bencherian** [comment]
Sorry, I just realized the code pushed to this branch is broken...I'll update this shortly.

**Comment by bencherian** [comment]
Ok, this should actually run and pass tests now.

**Comment by bencherian** [comment]
Would it be possible to get some feedback on whether this PR might be merged?
